### PR TITLE
ENH: add validate as a param to join (#46622)

### LINKED ISSUE #46622: ENH: Join - Add a parameter to check for duplicates
#### Is your feature request related to a problem?

It is arguably more common than not to join dataframes on a unique index. Optional check for uniqueness would help to prevent hidden errors and time spent on debugging.

#### Describe the solution you'd like

`DataFrame.join` should get a new parameter `on_unique` that would, if set to True, check if the index/columns on which joining is performed have duplicates and raise error if yes. Default should be False to keep backward compatibility.

#### API breaking implications

Default set to False ensures no issues with compatibility.

### PR DESCRIPTION
**Pull request by gaotian98**
- [ ] closes #46622

- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature

- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/frame/methods/test_join.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_join.py:110]: Could you split this into multiple tests so it's easier to see which conditions are being tested?

File: pandas/tests/frame/methods/test_join.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_join.py:110]: Is this test similar to the `_correctness` tests? Maybe can be combined with `@pytest.mark.parameterize`?

Comment (mroeschke) [reply pandas/tests/frame/methods/test_join.py:110]: Sure, that's great.

**Review by mroeschke — APPROVED** [review]

**Comment by gaotian98** [comment]
Does it matter for the change to be merged if one code check fails? The failure seems to come from a file I never touched, while the previous two commits were able to pass all checks.

**Comment by mroeschke** [comment]
> Does it matter for the change to be merged if one code check fails?

This was a known failure affecting all builds. Just waiting on if other maintainers would like to review

**Comment by gaotian98** [comment]
> > Does it matter for the change to be merged if one code check fails?
> 
> This was a known failure affecting all builds. Just waiting on if other maintainers would like to review

Sure. Thank you!

File: pandas/tests/frame/methods/test_join.py
Comment (phofl) [inline pandas/tests/frame/methods/test_join.py:111]: Could you either try parametrizing this or creating multiple tests? This is hard to debug if something goes wrong

Comment (gaotian98) [reply pandas/tests/frame/methods/test_join.py:111]: If this test fails, it's easy to trace the issue with the line and corresponding comments in this function. I don't think of an easy way of parametrizing that does not make this messier. Could you double check?

Comment (phofl) [reply pandas/tests/frame/methods/test_join.py:111]: Multiple tests are fine too. 

If the first fails all other checks are not executed. Also it is not easy to see how the df looks if a check further below fails

Comment (gaotian98) [reply pandas/tests/frame/methods/test_join.py:111]: Just changed. Does it look better now? Thanks.

Comment (phofl) [reply pandas/tests/frame/methods/test_join.py:111]: This is still one big test?

Comment (gaotian98) [reply pandas/tests/frame/methods/test_join.py]: Oh sorry. Forgot to add test_join.py after resolving conflict with the upstream/main. Please check again

**Review by mroeschke — APPROVED** [review]

**Review by mroeschke — COMMENTED** [review]
LGTM. Looks good to merge pending another look from @phofl

File: pandas/core/frame.py
Comment (jreback) [inline pandas/core/frame.py:9438]: can you add a versionadded 1.5 here

File: pandas/tests/frame/methods/test_join.py
Comment (jreback) [inline pandas/tests/frame/methods/test_join.py:110]: can you put these at the top of the file (the fixtures)

File: pandas/core/frame.py
Comment (jreback) [inline pandas/core/frame.py:9437]: if you could add an example using this would be great as well.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm some comments, ping on green.

File: pandas/core/frame.py
Comment (gaotian98) [reply pandas/core/frame.py:9438]: done, thanks!

File: pandas/tests/frame/methods/test_join.py
Comment (gaotian98) [reply pandas/tests/frame/methods/test_join.py]: done!

File: pandas/core/frame.py
Comment (gaotian98) [reply pandas/core/frame.py:9437]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good, it might be worth adding something to https://pandas.pydata.org/pandas-docs/dev/user_guide/merging.html?highlight=join#joining-on-index (there is a lot of sections, so this is a *maybe*).

**Comment by jreback** [comment]
thanks @gaotian98 very nice!

**Pull request merged**
